File: src/mod11_gringotts_concurrency/lesson7_exchanger/java/Main.java
package mod11_gringotts_concurrency.lesson7_exchanger.java;


import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.Exchanger;

public class Main {
    public static void main(String[] args) {

        BankAccount ollivander = new BankAccount("Ollivander's account", 15_000);
        BankAccount hagrid = new BankAccount("Hadrid's account", 500_000);

        Exchanger<String> exchanger=new Exchanger<>();


        CyclicBarrier cyclicBarrier =
                new CyclicBarrier(2, () ->
                        System.out.println("Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
                        ollivander.getAccountName() + ":" + ollivander.getBalance()));

        Thread thread1 = new Thread1(hagrid, ollivander,cyclicBarrier,exchanger);
        Thread thread2 = new Thread(new RunnableForThread2(hagrid, ollivander,cyclicBarrier,exchanger));

        thread1.start();
        thread2.start();

    }
}

File: src/mod11_gringotts_concurrency/lesson7_exchanger/kotlin/Main.kt
package mod11_gringotts_concurrency.lesson7_exchanger.kotlin

import java.util.concurrent.CyclicBarrier
import java.util.concurrent.Exchanger
import kotlin.concurrent.thread

fun main() {

    val hagridAccount = BankAccount("Hadrid's account", 500_000)
    val ollivanderAccount = BankAccount("Ollivander's account", 15_000)
    val cyclicBarrier = CyclicBarrier(2) {
        println(
            "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
                    ollivanderAccount.accountName + ":" + ollivanderAccount.balance
        )
    }

    val exchanger = Exchanger<String>()
    val thread1 = thread {
        for (i in 0 until 100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
            if (i == 60) {
                val response = exchanger.exchange("Good job! Now you can continue to work without haste.");
                println("Hadrid received a msg: $response")
            }
        }
        cyclicBarrier.await()
    }
    val thread2 = thread {
        for (i in 0 until 100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
            if (i == 60) {
                val response = exchanger.exchange("I have completed the ${i}th transaction.")
                println("Student received msg:$response")
            }
        }
        cyclicBarrier.await()
    }


}